// devexpress-reporting/designer/wizard/reportWizardState.js
/**
* DevExpress HTML/JS Reporting (designer\wizard\reportWizardState.js)
* Version:  20.2.5
* Build date: Jan 18, 2021
* Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xrReport_1 = require("../controls/metadata/xrReport");
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var $ = require("jquery");
var ReportType;
(function (ReportType) {
    ReportType[ReportType["Empty"] = 3] = "Empty";
    ReportType[ReportType["Databound"] = 0] = "Databound";
    ReportType[ReportType["Vertical"] = 1] = "Vertical";
    ReportType[ReportType["Label"] = 2] = "Label";
})(ReportType = exports.ReportType || (exports.ReportType = {}));
var GraphicsUnit;
(function (GraphicsUnit) {
    GraphicsUnit[GraphicsUnit["World"] = 0] = "World";
    GraphicsUnit[GraphicsUnit["Display"] = 1] = "Display";
    GraphicsUnit[GraphicsUnit["Pixel"] = 2] = "Pixel";
    GraphicsUnit[GraphicsUnit["Point"] = 3] = "Point";
    GraphicsUnit[GraphicsUnit["Inch"] = 4] = "Inch";
    GraphicsUnit[GraphicsUnit["Document"] = 5] = "Document";
    GraphicsUnit[GraphicsUnit["Millimeter"] = 6] = "Millimeter";
})(GraphicsUnit = exports.GraphicsUnit || (exports.GraphicsUnit = {}));
exports.defaultPageSetupState = {
    paperKind: xrReport_1.paperKind.defaultVal,
    landscape: false,
    marginTop: 1,
    marginRight: 1,
    marginBottom: 1,
    marginLeft: 1,
    width: analytics_internal_1.papperKindMapper[xrReport_1.paperKind.defaultVal].width / 100,
    height: analytics_internal_1.papperKindMapper[xrReport_1.paperKind.defaultVal].height / 100,
    unit: GraphicsUnit.Inch
};
exports.defaultReportWizardState = {
    masterDetailInfoCollection: [],
    pageSetup: $.extend(true, {}, exports.defaultPageSetupState),
    colorScheme: {},
    ignoreNullValuesForSummary: false
};
